### ISSUE TITLE
A peft model requires .train() to be called in order to store gradients during backprop

### ISSUE THREAD

**Issue by edbeeching** [huggingface/peft#120]
A peft model requires .train() to be called in order to store gradients during backprop.
cc @younesbelkada 
peft version 0.2.0.dev0

See the example here:
```
import torch
from transformers import AutoModelForCausalLM
from peft import LoraConfig, get_peft_model, prepare_model_for_int8_training


model = AutoModelForCausalLM.from_pretrained("edbeeching/gpt-neo-125M-imdb", load_in_8bit=True, device_map="auto")

lora_config = LoraConfig(
    r=16,
    lora_alpha=32,
    target_modules=["k_proj", "v_proj", "q_proj", "out_proj"],  # Are these the correct layers to target with LoRA ?
    lora_dropout=0.05,
    bias="none",
    task_type="CAUSAL_LM",
)
prepare_model_for_int8_training(model)
model = get_peft_model(model, lora_config)

with torch.cuda.amp.autocast():  
    gen = model.generate(input_ids=torch.LongTensor([0,1,2,3]).unsqueeze(0)) 
    # model.train() # <-- UNCOMMENT THIS LINE
    output = model(torch.LongTensor([0,1,2,3]).to("cuda"))
loss = output[0].mean()
loss.backward()
for n,p in model.named_parameters(): 
    print(n, p.grad)
```

**Comment by younesbelkada** [comment]
@pacman100 wdyt of adding `model.train()` in `get_peft_model`, happy to have a look and open a PR!

**Comment by pacman100** [comment]
Hello @edbeeching, aren't all models by default in `eval` mode and we need to set `model.train()` at the start of the training loop? Or am I missing something?

**Comment by younesbelkada** [comment]
I think you are right, maybe it should be the user's responsibility to add `.train()` when doing training
What about adding it inside `prepare_int8_model_for_training` ?

**Comment by edbeeching** [comment]
I completely agree that we _should_ call model.train() at the start of the training loop. However in this example if you remove these lines:
```
prepare_model_for_int8_training(model)
model = get_peft_model(model, lora_config)
```
Gradients are still calculated on a standard transformers model, even when in eval model.

**Comment by pacman100** [comment]
Hello, the issue was `with torch.cuda.amp.autocast(): `. Below code works without `model.train()` to get the gradients:
```diff
import torch
from transformers import AutoModelForCausalLM
from peft import LoraConfig, get_peft_model, prepare_model_for_int8_training


model = AutoModelForCausalLM.from_pretrained("edbeeching/gpt-neo-125M-imdb", load_in_8bit=True, device_map="auto")

lora_config = LoraConfig(
    r=16,
    lora_alpha=32,
    target_modules=["k_proj", "v_proj", "q_proj", "out_proj"],  # Are these the correct layers to target with LoRA ?
    lora_dropout=0.05,
    bias="none",
    task_type="CAUSAL_LM",
)
prepare_model_for_int8_training(model)
model = get_peft_model(model, lora_config)

- with torch.cuda.amp.autocast():  
gen = model.generate(input_ids=torch.LongTensor([0,1,2,3]).unsqueeze(0)) 
# model.train() # <-- UNCOMMENT THIS LINE
output = model(torch.LongTensor([0,1,2,3]).to("cuda"))
loss = output[0].mean()
loss.backward()
for n,p in model.named_parameters(): 
    print(n, p.grad)
```

**Comment by edbeeching** [comment]
A good to know, thanks!
